iree/compiler/Dialect/Flow/IR/FlowOps.cpp

```diff
@@ -263,6 +265,118 @@ static LogicalResult verifyVariableStoreIndirectOp(
 // flow.dispatch.region
 //===----------------------------------------------------------------------===//
 
+llvm::Optional<std::pair<DispatchRegionOp, Operation *>>
+DispatchRegionOp::formFromAnchorOp(Value workload, Operation *anchorOp,
+                                   OpBuilder &builder) {
+  builder.setInsertionPoint(anchorOp);
+  auto loc = anchorOp->getLoc();
+  // Map anchor into new dispatch region.
+  llvm::SmallVector<Value, 4> capturedInputs(anchorOp->getOperands());
+  llvm::SmallVector<Type, 1> types(anchorOp->getResultTypes().begin(),
+                                   anchorOp->getResultTypes().end());
+  auto drOp =
+      builder.create<DispatchRegionOp>(loc, types, workload, capturedInputs);
+  auto *drBlock = new Block();
+  drOp.body().push_back(drBlock);
+  BlockAndValueMapping mapping;
+  for (Value capturedInput : capturedInputs) {
+    auto blockArg = drBlock->addArgument(capturedInput.getType());
+    mapping.map(capturedInput, blockArg);
+  }
+
+  // Create new body.
+  builder.setInsertionPointToEnd(drBlock);
+  auto *newAnchorOp = builder.clone(*anchorOp, mapping);
+  builder.create<IREE::Flow::ReturnOp>(loc, newAnchorOp->getResults());
+
+  // Replace anchor uses with region result.
+  for (auto it : llvm::enumerate(anchorOp->getResults())) {
+    it.value().replaceAllUsesWith(drOp.getResult(it.index()));
+  }
+  anchorOp->erase();
+  return std::make_pair(drOp, newAnchorOp);
+}
+
+void DispatchRegionOp::dceOperandsAndResults(DispatchRegionOp &op) {
+  OpBuilder builder(op.getContext());
+  ClosureOpDce dce(op, op.body().front(), /*variadicOffset=*/1);
+  op = llvm::cast<DispatchRegionOp>(dce.optimize(builder));
+}
+
+ResultRange DispatchRegionOp::appendResults(
+    DispatchRegionOp &self, llvm::SmallVectorImpl<Value> &addlResults,
+    OpBuilder &builder) {
+  Block &block = self.body().front();
+
+  unsigned origNumResults = self.getNumResults();
+  llvm::SmallVector<Type, 4> newTypes(self.getResultTypes().begin(),
```

Comment:
to_vector here as well; then you could do something like:
```auto newTypes = llvm::to_vector<4>(llvm::concat(self.getResultTypes(), newTypes));```

(and below with `returns` too could just be `setOperands(llvm::concat(terminator->getOperands(), addlResults))`)